src/engine/renderer/gl_shader.cpp

```diff
@@ -770,9 +770,29 @@ std::string     GLShaderManager::BuildGPUShaderText( Str::StringRef mainShaderNa
 	// so we have to reset the line counting.
 	env += "#line 0\n";
 
-	std::string shaderText = env + libs + GetShaderText(filename);
+	std::string shaderText = env + libs + GetShaderText( filename );
 
-	return shaderText;
+	std::istringstream shaderTextStream( shaderText );
+	std::string shaderMain;
+
+	std::string line;
+
+	while ( std::getline( shaderTextStream, line, '\n' ) ) {
+		std::string::size_type position = line.find( "#insert" );
+		if ( position == std::string::npos ) {
+			shaderMain += line + "\n";
+			continue;
+		}
+
+		std::string shaderInsertPath = line.substr( position + 8, std::string::npos );
+		if ( shaderType == GL_VERTEX_SHADER ) {
+			shaderMain += GetShaderText( "glsl/" + shaderInsertPath + "_vp.glsl" );
+		} else {
+			shaderMain += GetShaderText( "glsl/" + shaderInsertPath + "_fp.glsl" );
+		}
+	}
```

Comment:
Would it be better to just do:

```
#define __unv_include(x) #x
#define _unv_include(x) __unv_include( x ## __GLSL_SUFFIX )
#define unv_include(x) _unv_include(x)
```

Then, we just have to define __GLSL_SUFFIX as `_fp.glsl` or `_vp.glsl` in the code and we can let the C preprocessor do all the work instead of us.

Reply:
If we go this route I would recommand a `daemon` suffix than an `unv` one.

Reply:
Do `#include` works in our GLSL system to begin with?

Reply:
Never mind, I don't think it does. Feel free to ignore.

Reply:
It doesn't, the only other directives we use are ones that are processed by GLSL compilers. Anyway, I was just doing whatever the rest of gl_shaders.cpp was doing lol.

Reply:
Let's not have magic suffixes at all, that's just confusing. Just write the whole filename including the extension. Plus we may want to share a function between vertex and fragment shaders.

Reply:
I've removed the fixed suffix.